File: packages/sdk-node/src/resources/subscribers.ts
import type {
  CreateSubscriberRequest,
  UpdateSubscriberRequest,
  SubscriberResponse,
  ListSubscribersParams,
  RegisterPushTokenRequest,
  PushTokenResponse,
  PreferencesResponse,
  RequestOptions,
} from "../types";
import {
  SubscriberResponseSchema,
  PushTokenResponseSchema,
  toSubscriberResponse,
  toPushTokenResponse,
} from "../types";

type RequestFn = <T>(
  method: string,
  path: string,
  body?: Record<string, unknown>,
  options?: RequestOptions,
  parseResponse?: (raw: unknown) => T,
) => Promise<T>;

export class SubscribersResource {
  constructor(private request: RequestFn) {}

  async create(params: CreateSubscriberRequest): Promise<SubscriberResponse> {
    return this.request(
      "POST",
      "/subscribers",
      params as unknown as Record<string, unknown>,
      undefined,
      (raw) => toSubscriberResponse(SubscriberResponseSchema.parse(raw)),
    );
  }

  async list(params?: ListSubscribersParams): Promise<SubscriberResponse[]> {
    const query = new URLSearchParams();
    if (params?.limit) query.set("limit", String(params.limit));
    if (params?.offset) query.set("offset", String(params.offset));
    const qs = query.toString();
    const path = qs ? `/subscribers?${qs}` : "/subscribers";
    return this.request("GET", path, undefined, undefined, (raw) => {
      const arr = Array.isArray(raw) ? raw : [];
      return arr.map((r: unknown) => toSubscriberResponse(SubscriberResponseSchema.parse(r)));
    });
  }

  async get(subscriberId: string): Promise<SubscriberResponse> {
    return this.request(
      "GET",
      `/subscribers/${encodeURIComponent(subscriberId)}`,
      undefined,
      undefined,
      (raw) => toSubscriberResponse(SubscriberResponseSchema.parse(raw)),
    );
  }

  async update(subscriberId: string, params: UpdateSubscriberRequest): Promise<SubscriberResponse> {
    return this.request(
      "PATCH",
      `/subscribers/${encodeURIComponent(subscriberId)}`,
      params as unknown as Record<string, unknown>,
      undefined,
      (raw) => toSubscriberResponse(SubscriberResponseSchema.parse(raw)),
    );
  }

  async delete(subscriberId: string): Promise<void> {
    return this.request("DELETE", `/subscribers/${encodeURIComponent(subscriberId)}`);
  }

  async getPreferences(subscriberId: string): Promise<PreferencesResponse> {
    return this.request("GET", `/subscribers/${encodeURIComponent(subscriberId)}/preferences`);
  }

  async updatePreferences(subscriberId: string, params: Record<string, unknown>): Promise<PreferencesResponse> {
    return this.request("PATCH", `/subscribers/${encodeURIComponent(subscriberId)}/preferences`, params);
  }

  async registerPushToken(subscriberId: string, params: RegisterPushTokenRequest): Promise<PushTokenResponse[]> {
    return this.request(
      "POST",
      `/subscribers/${encodeURIComponent(subscriberId)}/push-tokens`,
      params as unknown as Record<string, unknown>,
      undefined,
      (raw) => {
        const arr = Array.isArray(raw) ? raw : [];
        return arr.map((r: unknown) => toPushTokenResponse(PushTokenResponseSchema.parse(r)));
      },
    );
  }

  async listPushTokens(subscriberId: string): Promise<PushTokenResponse[]> {
    return this.request(
      "GET",
      `/subscribers/${encodeURIComponent(subscriberId)}/push-tokens`,
      undefined,
      undefined,
      (raw) => {
        const arr = Array.isArray(raw) ? raw : [];
        return arr.map((r: unknown) => toPushTokenResponse(PushTokenResponseSchema.parse(r)));
      },
    );
  }

  async removePushToken(subscriberId: string, token: string): Promise<PushTokenResponse[]> {
    return this.request(
      "DELETE",
      `/subscribers/${encodeURIComponent(subscriberId)}/push-tokens/${encodeURIComponent(token)}`,
      undefined,
      undefined,
      (raw) => {
        const arr = Array.isArray(raw) ? raw : [];
        return arr.map((r: unknown) => toPushTokenResponse(PushTokenResponseSchema.parse(r)));
      },
    );
  }
}

File: packages/sdk-python/src/alrt_sdk/resources/subscribers.py
"""Subscribers resource — CRUD, preferences, push tokens."""
from __future__ import annotations

from typing import Any, Callable, Awaitable
from urllib.parse import quote

from alrt_sdk.types import (
    CreateSubscriberRequest,
    UpdateSubscriberRequest,
    SubscriberResponse,
    RegisterPushTokenRequest,
    PushTokenResponse,
    PreferencesResponse,
)


class SubscribersResource:
    def __init__(self, request: Callable[..., Any]) -> None:
        self._request = request

    def create(self, **kwargs: Any) -> SubscriberResponse:
        params = CreateSubscriberRequest(**kwargs)
        body = params.model_dump(by_alias=True, exclude_none=True)
        raw = self._request("POST", "/subscribers", json=body)
        return SubscriberResponse.model_validate(raw)

    def list(self, limit: int = 20, offset: int = 0) -> list[SubscriberResponse]:
        raw = self._request("GET", f"/subscribers?limit={limit}&offset={offset}")
        items = raw if isinstance(raw, list) else []
        return [SubscriberResponse.model_validate(r) for r in items]

    def get(self, subscriber_id: str) -> SubscriberResponse:
        raw = self._request("GET", f"/subscribers/{quote(subscriber_id)}")
        return SubscriberResponse.model_validate(raw)

    def update(self, subscriber_id: str, **kwargs: Any) -> SubscriberResponse:
        params = UpdateSubscriberRequest(**kwargs)
        body = params.model_dump(by_alias=True, exclude_none=True)
        raw = self._request("PATCH", f"/subscribers/{quote(subscriber_id)}", json=body)
        return SubscriberResponse.model_validate(raw)

    def delete(self, subscriber_id: str) -> None:
        self._request("DELETE", f"/subscribers/{quote(subscriber_id)}")

    def get_preferences(self, subscriber_id: str) -> PreferencesResponse:
        raw = self._request("GET", f"/subscribers/{quote(subscriber_id)}/preferences")
        return PreferencesResponse.model_validate(raw)

    def update_preferences(self, subscriber_id: str, preferences: dict) -> PreferencesResponse:
        raw = self._request("PATCH", f"/subscribers/{quote(subscriber_id)}/preferences", json=preferences)
        return PreferencesResponse.model_validate(raw)

    def register_push_token(self, subscriber_id: str, **kwargs: Any) -> list[PushTokenResponse]:
        params = RegisterPushTokenRequest(**kwargs)
        body = params.model_dump(by_alias=True, exclude_none=True)
        raw = self._request("POST", f"/subscribers/{quote(subscriber_id)}/push-tokens", json=body)
        items = raw if isinstance(raw, list) else []
        return [PushTokenResponse.model_validate(r) for r in items]

    def list_push_tokens(self, subscriber_id: str) -> list[PushTokenResponse]:
        raw = self._request("GET", f"/subscribers/{quote(subscriber_id)}/push-tokens")
        items = raw if isinstance(raw, list) else []
        return [PushTokenResponse.model_validate(r) for r in items]

    def remove_push_token(self, subscriber_id: str, token: str) -> list[PushTokenResponse]:
        raw = self._request("DELETE", f"/subscribers/{quote(subscriber_id)}/push-tokens/{quote(token)}")
        items = raw if isinstance(raw, list) else []
        return [PushTokenResponse.model_validate(r) for r in items]


class AsyncSubscribersResource:
    def __init__(self, request: Callable[..., Awaitable[Any]]) -> None:
        self._request = request

    async def create(self, **kwargs: Any) -> SubscriberResponse:
        params = CreateSubscriberRequest(**kwargs)
        body = params.model_dump(by_alias=True, exclude_none=True)
        raw = await self._request("POST", "/subscribers", json=body)
        return SubscriberResponse.model_validate(raw)

    async def list(self, limit: int = 20, offset: int = 0) -> list[SubscriberResponse]:
        raw = await self._request("GET", f"/subscribers?limit={limit}&offset={offset}")
        items = raw if isinstance(raw, list) else []
        return [SubscriberResponse.model_validate(r) for r in items]

    async def get(self, subscriber_id: str) -> SubscriberResponse:
        raw = await self._request("GET", f"/subscribers/{quote(subscriber_id)}")
        return SubscriberResponse.model_validate(raw)

    async def update(self, subscriber_id: str, **kwargs: Any) -> SubscriberResponse:
        params = UpdateSubscriberRequest(**kwargs)
        body = params.model_dump(by_alias=True, exclude_none=True)
        raw = await self._request("PATCH", f"/subscribers/{quote(subscriber_id)}", json=body)
        return SubscriberResponse.model_validate(raw)

    async def delete(self, subscriber_id: str) -> None:
        await self._request("DELETE", f"/subscribers/{quote(subscriber_id)}")

    async def get_preferences(self, subscriber_id: str) -> PreferencesResponse:
        raw = await self._request("GET", f"/subscribers/{quote(subscriber_id)}/preferences")
        return PreferencesResponse.model_validate(raw)

    async def update_preferences(self, subscriber_id: str, preferences: dict) -> PreferencesResponse:
        raw = await self._request("PATCH", f"/subscribers/{quote(subscriber_id)}/preferences", json=preferences)
        return PreferencesResponse.model_validate(raw)

    async def register_push_token(self, subscriber_id: str, **kwargs: Any) -> list[PushTokenResponse]:
        params = RegisterPushTokenRequest(**kwargs)
        body = params.model_dump(by_alias=True, exclude_none=True)
        raw = await self._request("POST", f"/subscribers/{quote(subscriber_id)}/push-tokens", json=body)
        items = raw if isinstance(raw, list) else []
        return [PushTokenResponse.model_validate(r) for r in items]

    async def list_push_tokens(self, subscriber_id: str) -> list[PushTokenResponse]:
        raw = await self._request("GET", f"/subscribers/{quote(subscriber_id)}/push-tokens")
        items = raw if isinstance(raw, list) else []
        return [PushTokenResponse.model_validate(r) for r in items]

    async def remove_push_token(self, subscriber_id: str, token: str) -> list[PushTokenResponse]:
        raw = await self._request("DELETE", f"/subscribers/{quote(subscriber_id)}/push-tokens/{quote(token)}")
        items = raw if isinstance(raw, list) else []
        return [PushTokenResponse.model_validate(r) for r in items]
